### PR TITLE
chore: download browsers out of process

### PR DESCRIPTION
**Pull request by mxschmitt**
https://github.com/microsoft/playwright/issues/17394

### CODE REVIEW

File: packages/playwright-core/src/server/registry/download.ts [hunk 1]
```diff
@@ -15,68 +15,38 @@
...
+function downloadFileOutOfProcess(url: string, destinationPath: string, options: DownloadFileOptions = {}): Promise<{ error: Error | null }> {
+  const cp = childProcess.fork(path.join(__dirname, 'downloadMain.js'), [url, destinationPath, options.progressBarName || '', options.userAgent || '']);
+  const promise = new ManualPromise<{ error: Error | null }>();
+  cp.on('message', (message: any) => {
+    if (message?.method === 'log')
+      options.log?.(message.params.message);
+  });
+  cp.on('exit', code => {
+    if (code !== 0) {
+      promise.resolve({ error: new Error(`Download process exited with code ${code}`) });
```
Comment (dgozman) [inline packages/playwright-core/src/server/registry/download.ts:38]:
```suggestion
      promise.resolve({ error: new Error(`Download failure, code=${code}`) });
```

File: packages/playwright-core/src/server/registry/download.ts [hunk 2]
```diff
@@ -15,68 +15,38 @@
...
-    totalBytes = parseInt(response.headers['content-length'] || '0', 10);
-    log(`-- total bytes: ${totalBytes}`);
-    if (progressCallback)
-      response.on('data', onData);
-  }, (error: any) => fulfill({ error }));
+    // Node.js has a bug where the process can exit with 0 code even though there was an uncaught exception.
+    // Thats why we manually check if the destination file exists.
+    // https://github.com/microsoft/playwright/issues/17394
+    if (!fs.existsSync(destinationPath))
+      promise.resolve({ error: new Error(`Download process failed to write to ${destinationPath}`) });
```
Comment (dgozman) [inline packages/playwright-core/src/server/registry/download.ts:45]:
```suggestion
      promise.resolve({ error: new Error(`Download failure, ${destinationPath} does not exist`) });
```

Comment (dgozman) [inline packages/playwright-core/src/server/registry/download.ts:45]: Will this trigger a retry?

File: packages/playwright-core/src/server/registry/download.ts [hunk 3]
```diff
@@ -15,68 +15,38 @@
...
-        const error = new Error(`Download failed: server returned code ${response.statusCode} body '${content}'. URL: ${url}`);
-        // consume response data to free up memory
-        response.resume();
-        fulfill({ error });
-      };
-      response
-          .on('data', chunk => content += chunk)
-          .on('end', handleError)
-          .on('error', handleError);
+function downloadFileOutOfProcess(url: string, destinationPath: string, options: DownloadFileOptions = {}): Promise<{ error: Error | null }> {
```
Comment (dgozman) [inline packages/playwright-core/src/server/registry/download.ts:29]: Let's add a comment that links to the issue.

File: packages/playwright-core/src/server/registry/download.ts [hunk 4]
```diff
@@ -15,68 +15,38 @@
...
+  const cp = childProcess.fork(path.join(__dirname, 'downloadMain.js'), [url, destinationPath, options.progressBarName || '', options.userAgent || '']);
+  const promise = new ManualPromise<{ error: Error | null }>();
+  cp.on('message', (message: any) => {
+    if (message?.method === 'log')
+      options.log?.(message.params.message);
+  });
+  cp.on('exit', code => {
```
Comment (dgozman) [inline packages/playwright-core/src/server/registry/download.ts:36]: Please also handle `cp.on('error')`

File: packages/playwright-core/src/server/registry/download.ts (hunk 2 above)
Comment (mxschmitt) [reply packages/playwright-core/src/server/registry/download.ts]: Yes, see https://github.com/mxschmitt/playwright/blob/5284992135560cb7b0edcbf7f225a05d49b943a4/packages/playwright-core/src/server/registry/download.ts#L78

**Review by dgozman — APPROVED** [review]

**Pull request merged**